### PR TITLE
Fix Haptic glove transmission - GloveDevice

### PR DESCRIPTION
**Pull request by EhsanRanjbari**
See https://github.com/robotology/walking-teleoperation/issues/136#issuecomment-1916788820

### CODE REVIEW

File: msgs/thrift/WearableActuators.thrift
Comment (S-Dafarra) [inline msgs/thrift/WearableActuators.thrift:36]: The use of lists inside the message is not super clean, as you need to enforce in some way that each list has the same dimension (I guess).

What about defining a new message that instead contains a list of ``WearableActuatorCommand``?

I am thinking something like this:
```thrift
struct WearableActuatorCommand {
  1: ActuatorInfo info;
  2: double value;
  3: double duration;
}

struct WearableActuatorCommandList {
  1: list<WearableActuatorCommand> commands;
}
```

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:554]: Why this returns ``true`` if it does nothing? If it is not necessary anymore, consider removing it, or returning an error message

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:565]: Why not using a different vector?

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:311]: This is fishy. Why do you need the raw pointer to a private implementation?

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:127]: Why only the first?

**Review by S-Dafarra — COMMENTED** [review]
There a few things I did not manage to understand

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (EhsanRanjbari) [reply devices/HapticGlove/src/HapticGlove.cpp:554]: yes, you are right. These methods are no longer used. I will remove them.

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (EhsanRanjbari) [reply devices/HapticGlove/src/HapticGlove.cpp:311]: well that one is written in this way to be consistent with other parts of the code. For more details see the SenseGloveHapticActuator class.

Comment (S-Dafarra) [reply devices/HapticGlove/src/HapticGlove.cpp:311]: Still, it is fishy. Consider the private implementation as a key to your locker. You should be the only owner of your locker, but you are giving around copies of your key. It is not really safe.

**Comment by traversaro** [comment]
@EhsanRanjbari what is the status of this PR?

**Comment by EhsanRanjbari** [comment]
> @EhsanRanjbari what is the status of this PR?

Waiting to do some tests on the robot.

**Comment by S-Dafarra** [comment]
Hi @EhsanRanjbari, what is the status of this PR?

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (EhsanRanjbari) [reply devices/HapticGlove/src/HapticGlove.cpp:311]: An issue to discuss and fix this is opened: https://github.com/ami-iit/element_haptic-gloves/issues/192

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (EhsanRanjbari) [reply devices/HapticGlove/src/HapticGlove.cpp:566]: Well, because of the msg structure, I wanted to make it more readable.

File: msgs/thrift/WearableActuators.thrift
Comment (EhsanRanjbari) [reply msgs/thrift/WearableActuators.thrift]: Same as below, in this case, I think it is more readable since vibro and force are in separate vectors.

**Comment by EhsanRanjbari** [comment]
> Hi @EhsanRanjbari, what is the status of this PR?

Hi @S-Dafarra, well, I am still waiting to test them on the robot. unfortunately, both robots were unavailable most of the last week. However, preliminary test has been done with success (https://github.com/robotology/walking-teleoperation/issues/136)

**Comment by S-Dafarra** [comment]
Hi @EhsanRanjbari, did you manage to test on the robot?

**Comment by EhsanRanjbari** [comment]
> Hi @EhsanRanjbari, did you manage to test on the robot?

Hi. I have the robot today ftom 2. I will let you know how it went asap.

**Comment by EhsanRanjbari** [comment]
> > Hi @EhsanRanjbari, did you manage to test on the robot?
> 
> Hi. I have the robot today from 2. I will let you know how it went asap.

Hi again. I just finished testing and it went well. I also noticed that the vibration upon starting was not working. I fixed that by just aligning the branches with the remote. Now, everything looks fine according to this test.

**Comment by S-Dafarra** [comment]
Thanks @EhsanRanjbari! 

Nonetheless, I just realized that the changes to the ``WearableActuatorCommand`` can be problematic. First of all, it is weird that there is a single ``ActuatorInfo``, but then a vector of values (This is related to https://github.com/robotology/wearables/pull/204#discussion_r1474875228). For example, why for a single actuator of type ``HEATER`` should I receive a list of forces?
Secondly, it is not clear what I should use for the ``HEATER`` case.
Thirdly, this is probably breaking the multihaptic code, cc @dariosortino @Gianlucamilani 

At this point, I would suggest again not to edit the ``WearableActuatorCommand``, and instead add a new messages/methods to send and receive list of commands. This would preserve backward compatibility

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:556]: I would suggest to simply remove this method at this point

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:566]: Can you check that the size of ``forceValue`` and ``vibrotactileValue`` are correct?

File: interfaces/IWear/include/Wearable/IWear/Actuators/IHaptic.h
Comment (S-Dafarra) [inline interfaces/IWear/include/Wearable/IWear/Actuators/IHaptic.h:31]: Nitpick: it is more grammatically appropriate ``setHapticCommands``

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:81]: I think it should be ``gloveActuatorCommandInputPortName`` here

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:187]: You can simply use an ``if`` here and eventually throw an error

**Review by S-Dafarra — COMMENTED** [review]
Good move in adding a new message! Thanks for dealing with my comments @EhsanRanjbari. I have given it another round of review

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:68]: Maybe it is better to have a default instead here. In this way, old configuration files will not break.

Comment (EhsanRanjbari) [reply wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:68]: do you mean something like this:
```c++
    if (!config.check("gloveActuatorCommandInputPortName") || !config.find("gloveActuatorCommandInputPortName").isString()) {
        yWarning() << LogPrefix << "gloveActuatorCommandInputPortName parameter not found! Using the default one.";
        pImpl->gloveActuatorCommandInputPortName = "/WearableData/HapticGlove/RightHand/Actuators/input:i";
        return true;
    }
```

The problem here is that I need to find a way to specify the hand side part in the default port name.

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [reply devices/HapticGlove/src/HapticGlove.cpp:556]: Actually, I am wrong on this since it is an interface method. Just make sure that the message is not printed continuously in our use cases.

Comment (EhsanRanjbari) [reply devices/HapticGlove/src/HapticGlove.cpp:556]: > Actually, I am wrong on this since it is an interface method. Just make sure that the message is not printed continuously in our use cases.

Yes, actually I also tried to remove it back in time and eventually kept it for the same reason. Since it has not been called in glove devices the message is not being printed.

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (EhsanRanjbari) [reply wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp]: with https://github.com/robotology/wearables/pull/204/commits/12b55ac363c68d0e7445ca212bdd73a87b63d3ba I tried to specify the right and left hand's default actuator port name by finding the hand side parameter in the `IWearActuatorsWrapper.cpp`.

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:68]: This works only for the glove part, but this is used also for other devices. I don't think this can be used here

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:571]: This should be checked before accessing the vectors, otherwise you risk a segmentation fault

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (EhsanRanjbari) [reply wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:68]: Right. Then I will add a similar parameter to the wrapper part.

File: interfaces/IWear/include/Wearable/IWear/Actuators/IHaptic.h
Comment (S-Dafarra) [inline interfaces/IWear/include/Wearable/IWear/Actuators/IHaptic.h:31]: Since it is a setter, it is convenient to use ``const``
```suggestion
    virtual bool setHapticCommands(const std::vector<double>& forceValue, const std::vector<double>& vibrotactileValue) const = 0;
```

File: devices/HapticGlove/src/HapticGlove.cpp
Comment (S-Dafarra) [inline devices/HapticGlove/src/HapticGlove.cpp:559]:
```suggestion
    bool setHapticCommands(const std::vector<double>& forceValue, const std::vector<double>& vibrotactileValue) const override
```

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:100]: It is better to move this inside the if above, when you open the port

**Comment by EhsanRanjbari** [comment]
Thanks, @S-Dafarra for the reviews. I commit to the last change suggestions.

**Review by S-Dafarra — APPROVED** [review]
Looks good to me! Thanks a lot for keeping iterating on this. If you could do a final test on the robot, then this is good to go for me!

**Review by mebbaid — APPROVED** [review]

**Comment by EhsanRanjbari** [comment]
Today. I performed tests on the robot to verify this PR. The test was successful. Only, we noticed that not finding the parameter `` which is not used in the glove device, stops the haptic device. With this commit https://github.com/robotology/wearables/pull/204/commits/2bb61850b54d0ff3e7e27b52f2dd6bbc664e8e92 I tried to make it optional.

File: wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp
Comment (S-Dafarra) [inline wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:69]: The port is still opened, isn't it?

Comment (EhsanRanjbari) [reply wrappers/IWearActuators/src/IWearActuatorsWrapper.cpp:69]: Yes, fixed it with https://github.com/robotology/wearables/pull/204/commits/a6456185b90e688655dbdf331a7362e845ddc6e2

**Comment by S-Dafarra** [comment]
Thanks @EhsanRanjbari, looks good to me! @traversaro is there someone in particular to ask, or can I go ahead and merge this?

**Comment by EhsanRanjbari** [comment]
> @traversaro is there someone in particular to ask, or can I go ahead and merge this?

Just need to mention that this PR is highly dependent on this PR https://github.com/robotology/walking-teleoperation/pull/143

**Comment by S-Dafarra** [comment]
> > @traversaro is there someone in particular to ask, or can I go ahead and merge this?
> 
> Just need to mention that this PR is highly dependent on this PR [robotology/walking-teleoperation#143](https://github.com/robotology/walking-teleoperation/pull/143)

Actually, it is the opposite. This PR blocks https://github.com/robotology/walking-teleoperation/pull/143

**Comment by S-Dafarra** [comment]
Thinking about it @EhsanRanjbari can you bump the version from 1.8.0 to 1.9.0 in https://github.com/robotology/wearables/blob/91722ca62b65b4a2919fabf4ee58b3f90fd14a76/CMakeLists.txt#L7

In this way, we can change the minimum version required of ``wearables`` in ``walking-teleoperation``.

**Comment by EhsanRanjbari** [comment]
> Thinking about it @EhsanRanjbari can you bump the version from 1.8.0 to 1.9.0 in
> 
> https://github.com/robotology/wearables/blob/91722ca62b65b4a2919fabf4ee58b3f90fd14a76/CMakeLists.txt#L7
> 
> In this way, we can change the minimum version required of `wearables` in `walking-teleoperation`.

Done! https://github.com/robotology/wearables/pull/204/commits/32ff17c018a0015e9859a3ee1496cfd4329bc6d9

**Comment by traversaro** [comment]
> Thanks @EhsanRanjbari, looks good to me! @traversaro is there someone in particular to ask, or can I go ahead and merge this?

After Lorenzo Rapetti left, I think we do not have any obvious candidate for maintainer. Feel free to merge and release/tag, thanks!

**Comment by lrapetti** [comment]
> > Thanks @EhsanRanjbari, looks good to me! @traversaro is there someone in particular to ask, or can I go ahead and merge this?
> 
> After Lorenzo Rapetti left, I think we do not have any obvious candidate for maintainer. Feel free to merge and release/tag, thanks!

Probably before merging `CHANGELOG` could be updated as well :wink:

**Comment by S-Dafarra** [comment]
> Probably before merging `CHANGELOG` could be updated as well 😉

I think it could be a good time to follow https://github.com/robotology/idyntree/pull/1162 and start using the automatic generation of CHANGELOG of GitHub releases 🤔

**Comment by S-Dafarra** [comment]
> > Probably before merging `CHANGELOG` could be updated as well 😉
> 
> I think it could be a good time to follow [robotology/idyntree#1162](https://github.com/robotology/idyntree/pull/1162) and start using the automatic generation of CHANGELOG of GitHub releases 🤔

I will deal with this in a separate PR. For the moment, I am going to merge this. Thanks a lot @EhsanRanjbari

**Pull request merged**
